Deprecate torch.nn.utils.stateless.functional_call

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92280
* #92279
* #92278

This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

zou3519 committed:
Deprecate torch.nn.utils.stateless.functional_call

This PR:
- Updates the docs to say it is deprecated
- Raises a DeprecationWarning (which is surpressed by PyTorch, so most
users won't actually see it)
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a DeprecationWarning (which is surpressed by PyTorch, so most
users won't actually see it)
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

albanD commented on torch/nn/utils/stateless.py:
```diff
@@ -178,6 +179,12 @@ def functional_call(
 
+    .. warning::
+
+        This API is deprecated as of PyTorch 2.0.0 and will be removed in a future
```
2.0 is enough I think

albanD reviewed (approved):
SGTM

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Deprecate torch.nn.utils.stateless.functional_call"


This PR:
- Updates the docs to say it is deprecated
- Raises a UserWarning
- Changes most of the callsites inside PyTorch to use
torch.func.functional_call, minus the test_stateless testing.

The motivation behind this is that we can now align behind a single
functional_call API in PyTorch.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
